zvariant/src/signature.rs

```diff
@@ -511,6 +511,47 @@ impl std::fmt::Display for OwnedSignature {
     }
 }
 
+/// [`Signature`] that identifies a complete type.
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, Type)]
+pub struct CompleteType<'a>(Signature<'a>);
+
+assert_impl_all!(CompleteType<'_>: Send, Sync, Unpin);
+
+impl<'a> CompleteType<'a> {
+    /// Returns the underlying [`Signature`]
+    pub fn signature(&self) -> &Signature<'a> {
+        &self.0
+    }
+}
+
+impl<'a> Display for CompleteType<'a> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        std::fmt::Display::fmt(&self.0.as_str(), f)
+    }
+}
+
+impl<'a> TryFrom<Signature<'a>> for CompleteType<'a> {
```

Comment:
would be great to provide:

* `TryFrom<&str>` and `TryFrom<String>`.
* `Deref<Target = Signature>`.
* `From<Signature> for CompleteType`.

Reply:
they are not needed at this point.
"Deref should only be implemented for smart pointers to avoid confusion." iow, we shouldn't abuse it. Perhaps we do.

Reply:
From<Signature> can't be implemented, as it may throw errors. Use TryFrom.

Reply:
I wonder if Rust is not smart enough to figure out you can go try_from &str/String -> Signature -> CompleteType already (is it transitive?).

Reply:
> "Deref should only be implemented for smart pointers to avoid confusion."

Where does it say that?

> we shouldn't abuse it. Perhaps we do.

I don't think so. I think it's good for wrapper types like these.

> I wonder if Rust is not smart enough to figure out you can go try_from &str/String -> Signature -> CompleteType already (is it transitive?).

I doubt it. It's not always desired.

Reply:
> > I wonder if Rust is not smart enough to figure out you can go try_from &str/String -> Signature -> CompleteType already (is it transitive?).
> 
> I doubt it. It's not always desired.

Or maybe. I don't know. :shrug:

Reply:
Where does it say that? -> https://doc.rust-lang.org/std/ops/trait.Deref.html#

Reply:
I leave that nice to have out for now.